newlib/libc/machine/arc64/strlen.S

```diff
@@ -0,0 +1,216 @@
+/*
+   Copyright (c) 2021, Synopsys, Inc. All rights reserved.
+
+   Redistribution and use in source and binary forms, with or without
+   modification, are permitted provided that the following conditions are met:
+
+   1) Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+   2) Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+   3) Neither the name of the Synopsys, Inc., nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+   AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+   IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+   ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+   LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+   CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+   SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+   INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+   ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+   POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <sys/asm.h>
+
+; Code Brief:
+; Correct input pointer alignment and perform four 8 byte block search
+; This code is optimized for big strings, but also performs well for small
+; ones
+;
+; R0 const char* ptr (string to measure)
+; ret (R0):
+;		- unsigned (string size)
+;
+
+#if defined (__ARC64_ARCH64__)
+
+ENTRY (strlen)
+
+; If address is 4 byte aligned, we can just directly read 4 bytes
+bmsk.f	r3,		r0,		1
+beq.d			@.L_4B_step
+
+; Store r0 for size calculation when returning
+MOVP	r13,	r0
+
+subl	r3,		r3,		1		; [0]
+asl		r3,		r3,		2		; [1]
+
+; Jump depending on the alignment
+bi	    [r3]
+; Read 3 bytes
+ldb.ab	r10,	[r13,	+1]
+cmp		r10,	0
+beq 			@.L_return_size
```

Comment:
we have `breq` instruction which combines a `cmp` and  a `beq` instructions. Hence, your case block can be shorter.
Moreover, you can avoid all the jumps in the case blocks:
```asm
.case0:
ldb.ab r10,[r13,1]
.case1:
ldb.ab r11,[r13,1]
.case2:
ldb.ab r12,[r13,1]

cmp  r10,0
cmp.ne r11,0
cmp.ne  r12,0
beq  @.L_return_size
```
Or just skip all the alignment and go directly to a 64bit search

Reply:
I now understand that ARC can handle unaligned accesses well enough, and seeing the results are a lot better I will remove the alignment.
However, I have a question for you. I could indeed use breq to reduce block size to 2, but placing the jumps outside of it would make the size (found via r13) possibly wrong due to all the writebacks and I would need to know in which branch the byte was found.
Is there something I am not seeing, or would I need to encode it in a temporary register in the branches (like a bset in the switch block) and then find last set to adjust r13.